R/verb-joins.R

```diff
@@ -234,44 +234,94 @@ add_join <- function(x, y, type, by = NULL, sql_on = NULL, copy = FALSE,
   na_matches <- arg_match(na_matches, c("na", "never"), error_call = call)
 
-  vars <- join_vars(op_vars(x), op_vars(y), type = type, by = by, suffix = suffix, call = call)
+  # the table alias can only be determined after `select()` was inlined.
+  # This works even though `by` is used in `join_inline_select()` and updated
+  # because this does not touch `by$x_as` and `by$y_as`.
+  # TODO can this be really be done before inlining?
+  join_alias <- check_join_alias(x_as, y_as, sql_on, call)
 
   inline_result <- join_inline_select(x$lazy_query, by$x, by$on)
   x_lq <- inline_result$lq
   x_vars <- inline_result$vars
   by_x_org <- by$x
   by$x <- inline_result$by
 
+  new_query <- join_needs_new_query(x$lazy_query, join_alias, type)
+  if (new_query) {
+    x_join_vars <- tibble(
+      name = op_vars(x),
+      table = list(1L),
+      var = as.list(x_vars)
+    )
+    table_id <- 2L
+  } else {
+    x_join_vars <- x_lq$vars
+    table_id <- vctrs::vec_size(x_lq$table_names) + 1L
+  }
+
   inline_result <- join_inline_select(y$lazy_query, by$y, by$on)
   y_lq <- inline_result$lq
   y_vars <- inline_result$vars
   by$y <- inline_result$by
 
-  vars <- update_join_vars(vars, x_lq, y_lq, x_vars, y_vars, type, by)
-
-  # the table alias can only be determined after `select()` was inlined.
-  # This works even though `by` is used in `join_inline_select()` and updated
-  # because this does not touch `by$x_as` and `by$y_as`.
-  join_alias <- check_join_alias(x_as, y_as, sql_on, call)
+  y_join_vars <- tibble(
+    name = op_vars(y),
+    table = list(table_id),
+    var = as.list(y_vars)
+  )
 
-  x_name <- unclass(query_name(x_lq))
-  y_name <- unclass(query_name(y_lq))
-  by[c("x_as", "y_as")] <- join_two_table_alias(x_name, y_name, join_alias$x, join_alias$y)
-  by$x_as <- ident(by$x_as)
-  by$y_as <- ident(by$y_as)
+  vars <- multi_join_vars(
+    x_join_vars = x_join_vars,
+    y_join_vars = y_join_vars,
+    by_x_org = by_x_org,
+    by_y = by$y,
+    type = type,
+    table_id = table_id,
+    suffix = suffix
+  )
 
-  lazy_join_query(
-    x = x_lq,
-    y = y_lq,
-    vars = vars,
+  joins_data <- new_joins_data(
+    x_lq,
+    y_lq,
+    new_query = new_query,
     type = type,
     by = by,
-    suffix = suffix,
-    na_matches = na_matches,
-    call = call
+    na_matches = na_matches
+  )
+
+  table_names_y <- tibble(
+    as = join_alias$y %||% NA_character_,
+    name = as.character(query_name(y_lq) %||% NA)
   )
+
+  if (new_query) {
+    table_names_x <- tibble(
+      as = join_alias$x %||% NA_character_,
+      name = as.character(query_name(x_lq) %||% NA)
+    )
+
+    out <- lazy_multi_join_query(
+      x = x_lq,
+      joins = joins_data,
+      table_names = vctrs::vec_rbind(table_names_x, table_names_y),
+      vars = vars
+    )
+    return(out)
+  }
+
+  # `x_lq` must be a `lazy_multi_join_query` so it can be modified directly
+  if (!is_null(join_alias$x)) {
+    x_lq$table_names$as[[1]] <- join_alias$x
```

Comment:
When I worked through the example you suggested, I noticed that an `NA` was replaced by a `NULL` which seems mildly fishy.

Reply:
I can't quite follow where an `NA` was replaced by a `NULL`.

Reply:
IIRC, when I stepped through the code ` x_lq$table_names$as[[1]]` was `NA` and `join_alias$x` was `NULL`

Reply:
I refactored the table names logic and it is now a bit easier.